### ISSUE TITLE
Track representation switchs in adaptive video

### ISSUE THREAD

**Issue by emanresu98** [google/ExoPlayer#6096]
I am trying to track the audio and video representations switch when playing an adaptive video with Exoplayer.  I used the code implemented in the logger to track the events:

 ```
 @Override
  public void onTracksChanged(
      EventTime eventTime, TrackGroupArray ignored, TrackSelectionArray trackSelections) {
    MappedTrackInfo mappedTrackInfo =
        trackSelector != null ? trackSelector.getCurrentMappedTrackInfo() : null;
    if (mappedTrackInfo == null) {
      logd(eventTime, "tracksChanged", "[]");
      return;
    }
    logd("tracksChanged [" + getEventTimeString(eventTime) + ", ");
    // Log tracks associated to renderers.
    int rendererCount = mappedTrackInfo.getRendererCount();
    for (int rendererIndex = 0; rendererIndex < rendererCount; rendererIndex++) {
      TrackGroupArray rendererTrackGroups = mappedTrackInfo.getTrackGroups(rendererIndex);
      TrackSelection trackSelection = trackSelections.get(rendererIndex);
      if (rendererTrackGroups.length > 0) {
        logd("  Renderer:" + rendererIndex + " [");
        for (int groupIndex = 0; groupIndex < rendererTrackGroups.length; groupIndex++) {
          TrackGroup trackGroup = rendererTrackGroups.get(groupIndex);
          String adaptiveSupport =
              getAdaptiveSupportString(
                  trackGroup.length,
                  mappedTrackInfo.getAdaptiveSupport(rendererIndex, groupIndex, false));
          logd("    Group:" + groupIndex + ", adaptive_supported=" + adaptiveSupport + " [");
          for (int trackIndex = 0; trackIndex < trackGroup.length; trackIndex++) {
            String status = getTrackStatusString(trackSelection, trackGroup, trackIndex);
            String formatSupport =
                getFormatSupportString(
                    mappedTrackInfo.getTrackSupport(rendererIndex, groupIndex, trackIndex));
            logd(
                "      "
                    + status
                    + " Track:"
                    + trackIndex
                    + ", "
                    + Format.toLogString(trackGroup.getFormat(trackIndex))
                    + ", supported="
                    + formatSupport);
          }
          logd("    ]");
        }
        // Log metadata for at most one of the tracks selected for the renderer.
        if (trackSelection != null) {
          for (int selectionIndex = 0; selectionIndex < trackSelection.length(); selectionIndex++) {
            Metadata metadata = trackSelection.getFormat(selectionIndex).metadata;
            if (metadata != null) {
              logd("    Metadata [");
              printMetadata(metadata, "      ");
              logd("    ]");
              break;
            }
          }
        }
        logd("  ]");
      }
    }
    // Log tracks not associated with a renderer.
    TrackGroupArray unassociatedTrackGroups = mappedTrackInfo.getUnmappedTrackGroups();
    if (unassociatedTrackGroups.length > 0) {
      logd("  Renderer:None [");
      for (int groupIndex = 0; groupIndex < unassociatedTrackGroups.length; groupIndex++) {
        logd("    Group:" + groupIndex + " [");
        TrackGroup trackGroup = unassociatedTrackGroups.get(groupIndex);
        for (int trackIndex = 0; trackIndex < trackGroup.length; trackIndex++) {
          String status = getTrackStatusString(false);
          String formatSupport =
              getFormatSupportString(RendererCapabilities.FORMAT_UNSUPPORTED_TYPE);
          logd(
              "      "
                  + status
                  + " Track:"
                  + trackIndex
                  + ", "
                  + Format.toLogString(trackGroup.getFormat(trackIndex))
                  + ", supported="
                  + formatSupport);
        }
        logd("    ]");
      }
      logd("  ]");
    }
    logd("]");
  }
```

The code allows me to print the list of the available tracks as you explained in the documentation:

<img width="899" alt="Screenshot 2019-06-26 at 16 07 49" src="https://user-images.githubusercontent.com/52248590/60188982-7f1d7580-9830-11e9-9eae-857c3ccea2c2.png">

In this example, multiple video tracks are selected in renderer 0 groupe 1. How can I identify which track/representation is currently being used by the player?

The is no such api or method that allows to check the current representation more easily? something like `player.getCurrentVideoRepresentation` ?

Can you confirm that `onTracksChanged` will also be called when the representation is automatically changed by the player? In my test until now it was only called when I manually select a different track in the UI.

**Comment by ojw28** [comment]
When you say that you're interested in when the selected format is changed by the player, are you interested in:

1. The player decides to start buffering the different format
1. The player actually starts playing the different format

(1) will happen before (2), assuming that the adaptive switch occurs without any re-buffering taking place. You can implement `onLoadStarted` to figure out when the player changes the format that it's buffering. You can implement `onDownstreamFormatChanged` to figure out when the format that's actually being played changes.

**Comment by ojw28** [comment]
> Can you confirm that onTracksChanged will also be called when the representation is automatically changed by the player?

This doesn't happen. `onTracksChanged` is only called when the statically selected tracks (i.e. the ones that the player is allowed to switch between) change. Adding a `documentation candidate` label to this issue to better document this.

**Comment by stevemayhew** [comment]
@ojw28 because of the way `Format` is copied the `onDownstreamFormatChanged()` is not always logged (for example, when changing from 1080p@4Mbps to 1080p@8Mbps stream variants). 

Also @emanresu98 realize that this is not the currently playing variant format (as there likely are buffered parsed samples).

We would want:

1. All format changes to trigger events (even just bitrate)
2. Upstream format changed event also (this is currently only handled internally in the renderer)

Is this a bug or an RFE?

**Comment by ojw28** [comment]
> @ojw28 because of the way Format is copied the onDownstreamFormatChanged() is not always logged (for example, when changing from 1080p@4Mbps to 1080p@8Mbps stream variants).

I don't understand which copy you're referring to. It looks to me like it will always be logged, and this statement isn't specific enough allow me to determine otherwise. Please can be point to exactly where you believe this problem is occurring.

> Upstream format changed event also (this is currently only handled internally in the renderer)

What does upstream mean in this context? If it means "The player decides to start buffering the different format" as I mentioned above, you should use `onLoadStarted`.

Neither `onDownstreamFormatChanged` nor `onLoadStarted` are related to renderers. These events originate in the media source. The easiest way to receive these (and other) events is to register an `AnalyticsListener` using `SimpleExoPlayer.addAnalyticsListener`.

**Comment by stevemayhew** [comment]
@ojw28 Sorry Oliver, I misspoke the `Format` copy issue is different (`copyWithContainerInfo()` did not include the roles, that's in my [pull request](https://github.com/google/ExoPlayer/pull/6270/commits/21ea7d4f3509d84cacbcb76b8f09076adcf68517)). 

I mixed up up down/up stream. I was thinking furthest upstream are the samples closes to the playhead (as ExoPlayer's renderer feeds them to the codec), it's the reverse I see now.

So let me try again and see if this is clear/correct:

The format change access points are, listed in order from most distant to closest to the playhead:

 _upstreamFormat_ &mdash;  loaded  format (so the `onLoadStarted()` or `onLoadCompleted()` both give you the first indication a level shift occurred.
_downstreamFormat_ &mdash; most recent format read by the renderer from the SampleStream, as you said `MediaSourceEventListener#onDownstreamFormatChangedI()` signals this, media load data has the format.  Note this will give both audio and video format changes.
_videoInputFormat_ &mdash; VideoRendererEventListener#onVideoInputFormatChanged() is signaled for this, this is what is shown by the debug view in the demo (`Player.getVideoFormat()`)

In short there is no issue, and using the load event gives you when the actual level shift (if Format changes).

THanks
